Patchwork integration

ameba23 opened the pull request:
the branch has the features needed to add SSB identity recovery to patchwork;
- requests for forwarding shards
- attached encrypted blobs allowing for bigger secrets

both these features should be useful for other things outside this specific use case, but for testing its useful right now to have them in one branch

KGibb8 commented on forward-request/pull/for-own-shards.js:
```diff
@@ -0,0 +1,26 @@
+      // This query is specific to recovering SSB identities
+      pull.filter(shard => get(shard, 'value.content.attachment.name') === 'gossip.json'),
+      pull.unique(s => get(s, 'value.author')),
+      pull.map(shard => {
```
This won't work, calling back with the forward-requests from the collect callback expects an asynchronous action to follow, while returning the pull stream expects a synchronous action to follow. This needs to be `asyncMap` or `pullParamap`.

ameba23 replied:
:+1: good call

KGibb8 commented on forward-request/pull/from-others.js:
```diff
@@ -0,0 +1,29 @@
+    }, {
+      $filter: {
+        value: {
+          author: server.id
```
If its from others, shouldn't this be `$ne: server.id`?

ameba23 replied:
doh!

KGibb8 commented on forward-request/pull/from-self.js:
```diff
@@ -0,0 +1,29 @@
+    }, {
+      $filter: {
+        value: {
+          author: { $ne: server.id }
```
Inverse of the previous one, if its from self, shouldn't it be where the author is me? That said, I don't see where we'd use this.

KGibb8 replied:
Realising now, we will need `from-self` in the recovery when we're trying to find the relevant request for forward messages received.

KGibb8 commented on forward-request/pull/from-self.js:
```diff
@@ -0,0 +1,29 @@
+const { isForwardRequest } = require('ssb-dark-crystal-schema')
+
+module.exports = function (server) {
+  return function friends (opts = {}) {
```
incorrect name! :)

KGibb8 commented on forward-request/pull/from-others.js:
```diff
@@ -0,0 +1,29 @@
+const { isForwardRequest } = require('ssb-dark-crystal-schema')
+
+module.exports = function (server) {
+  return function friends (opts = {}) {
```
change!

KGibb8 commented on forward-request/pull/for-own-shards.js:
```diff
@@ -0,0 +1,26 @@
+    return pull(
+      shardsFromOthers(),
+      // This query is specific to recovering SSB identities
+      pull.filter(shard => get(shard, 'value.content.attachment.name') === 'gossip.json'),
```
Wondering if we can't dynamically pass in filter functions, that way we don't have make this so restrictive...

ameba23 replied:
ahh i like that idea

ameba23 replied:
@KGibb8 so i just changed it that it takes a filter function as an argument.  

probably it would be neater if this was a property of the `opts` object.  but i wasn't sure if `opts` has a standardised form and this would meddle with it.

KGibb8 commented on forward-request/pull/for-own-shards.js:
```diff
@@ -0,0 +1,26 @@
+      shardsFromOthers(),
+      // This query is specific to recovering SSB identities
+      pull.filter(shard => get(shard, 'value.content.attachment.name') === 'gossip.json'),
+      pull.unique(s => get(s, 'value.author')),
```
what exactly this is this doing?

KGibb8 commented on forward-request/pull/by-secret-owner.js:
```diff
@@ -0,0 +1,35 @@
+const { isFeedId } = require('ssb-ref')
+
+module.exports = function (server) {
+  return function bySecretOwner (secretOwner, opts = {}) {
```
this is `forSecretOwner` rather than `bySecretOwner`, because we're looking for all requests that have been created by person A, who may or may not be person B, and as person C, its a big semantic difference.

ameba23 replied:
im happy to change it.  i was imaging its short for 'indexed by secret owner', like we've got a query for shards indexed 'by root'. but if you are seeing it as 'forward requests authored by secret owner' then thats not what we want.

KGibb8 commented:
I'm feeling a strong opinion cropping up re the use of the `branch` field.

I've encountered a situation where a forward could possibly have two branch attributes, but there would be no way to determine the message type of those messages. Which makes for messy querying.

e.g. the `branch` attribute of a forward message _could reference the `forward-request` or it could reference the `shard` message.

This far, we haven't implemented having the shard key / id in the branch section. However if we were to do it in future, we'd be looking at significant code changes (more `server.get` / pull streams as queries) to determine the message type of the branch ID in question.

I'd propose instead of using the `branch` field, we go with SQL based convention, i.e. a forward schema changes from using the `branch` attribute to looking something like this:

```js
{ 
  type: 'dark-crystal/forward',
  requestId: '%.........',
  shardId: '%.........%',
  ...
}
```

ameba23 commented:
so i've ditched the branch field (we never used it anyway) and added shardId and requestId.

one thing im not sure of is how to handle multiple forward requests (as forward requests dont contain specific information about what they are requesting).  so as i've implemented it here, we will just take the first forward request the query gives as.  but thinking about it probably the most recent one is more relevant.

does it actually matter?  i guess the only thing we are using this for is to determine if we have an outstanding forward request that we didn't already respond to. so maybe we should have a forward-request query for a given secret owner which pairs them with your responses and only tells you about unanswered ones. we could call it `outstanding` which makes it sound rather great.

KGibb8 commented on forward/async/publish.js:
```diff
@@ -25,16 +27,26 @@ module.exports = function (server) {
-        })
+        pull(
+          pullForwardRequestsSecretOwner(get(shards[0], 'value.author')),
+          pull.collect((err, forwardRequests) => {
```
This change doesn't belong here, since its being used here in patchbay-darl-crystal - https://github.com/blockades/patchbay-dark-crystal/blob/master/views/forward/new.js#L99 - where we're not dealing with `forward-requests`. We either want to do this in a separate action i.e. `forward.async.publishRequestReply` or something like that. Or we make this forward publish function _not care about getting data beforehand_. I.e. it just simply formats the data (build) and publishes.

KGibb8 replied:
@ameba23 unless this accounts for its use elsewhere?

KGibb8 replied:
I've changed the Patchwork code to pass in only the `rootId` of the shard, so this should work for the time being. However, this is restricting as any forward responding to a request will always reference the _first one that was returned from the query_. See https://github.com/blockades/scuttle-dark-crystal/pull/54/files#diff-afc3167917ae15c1ebb29b31440b8908R36

KGibb8 commented on forward/async/publish.js:
```diff
@@ -25,16 +27,26 @@ module.exports = function (server) {
+            var requestId = null
+            if (err) return callback(err)
+            if (forwardRequests.length > 0) {
+              requestId = forwardRequests[0]
```
Bad. :)

KGibb8 replied:
We should instead be passing in the requestId, but that then buggers up the code in `patchbay-dark-crystal`. I'm much more in favour of a refactor, having an agnostic forward publish method that just takes params (much like `root.async.publish`, and then either getting the front-end to do the work, or having separate publish actions for the two different contexts we're now dealing with (with requests - Patchwork / without requests - Patchbay)

KGibb8 commented on forward/async/publish.js:
```diff
@@ -4,9 +4,11 @@ const get = require('lodash.get')
 module.exports = function (server) {
   const pullShardsByRoot = PullShardsByRoot(server)
+  const pullForwardRequestsSecretOwner = PullForwardRequestsSecretOwner(server)
   return function publishForward (root, recp, callback) {
```
It should be noted too that this breaks the pattern we use in the other publish functions, which expect a single `params` object, rather than two individual arguments.
